public/app/features/explore/state/utils.ts

```diff
@@ -119,12 +123,64 @@ export async function loadAndInitDatasource(
     }
   }
 
-  const historyKey = `grafana.explore.history.${instance.meta?.id}`;
-  const history = store.getObject<HistoryItem[]>(historyKey, []);
-  // Save last-used datasource
+  let history: HistoryItem[] = [];
+
+  const localStorageHistory = getLocalRichHistoryStorage();
+
+  const historyResults = await localStorageHistory.getRichHistory({
+    search: '',
+    sortOrder: SortOrder.Ascending,
+    datasourceFilters: [instance.name],
+    from: 0,
+    to: MAX_HISTORY_AUTOCOMPLETE_ITEMS,
+    starred: false,
+  });
+
+  // first, fill autocomplete with query history for that datasource
+  if ((historyResults.total || 0) > 0) {
+    historyResults.richHistory.forEach((historyResult: RichHistoryQuery) => {
+      historyResult.queries.forEach((q) => {
+        history.push({ ts: parseInt(historyResult.id, 10), query: q });
+      });
+    });
+  }
+
+  if (history.length < MAX_HISTORY_AUTOCOMPLETE_ITEMS) {
+    // check the last 100 mixed history results seperately
+    const historyMixedResults = await localStorageHistory.getRichHistory({
+      search: '',
+      sortOrder: SortOrder.Ascending,
+      datasourceFilters: [MIXED_DATASOURCE_NAME],
+      from: 0,
+      to: MAX_HISTORY_AUTOCOMPLETE_ITEMS,
```

Comment:
I think it's not well documented but `from` / `to` is used as a time filter where value is equal to "number of days since today" . `to: MAX_HISTORY_AUTOCOMPLETE_ITEMS` means getting queries that are at max 100 days old. Is `MAX_HISTORY_AUTOCOMPLETE_ITEMS` used just to have some arbitrary big number? If so I think it be more readable if it was a separate const or if we changed the API to return all items.

Reply:
So the to/from date filter is not required by the API, so I made it optional, then dont pass it in for local storage. The only place where it needs a time filter is the queries tab, which already has the default of going from now to the retention period of 2 weeks ([src](https://github.com/grafana/grafana/blob/main/public/app/features/explore/RichHistory/RichHistoryQueriesTab.tsx#L143)).